sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -131,6 +132,12 @@ class CacheManager extends Logging {
 
   /** Replaces segments of the given logical plan with cached versions where possible. */
   def useCachedData(plan: LogicalPlan): LogicalPlan = {
+    useCachedDataInternal(plan) transformAllExpressions {
+      case s: SubqueryExpression => s.withNewPlan(useCachedData(s.plan))
+    }
+  }
+
+  private def useCachedDataInternal(plan: LogicalPlan): LogicalPlan = {
```

Comment:
After rethinking about it, we do not need to add a new function. We can combine them into a single function, like:
```Scala
  /** Replaces segments of the given logical plan with cached versions where possible. */
  def useCachedData(plan: LogicalPlan): LogicalPlan = {
    val newPlan = plan transformDown {
      case currentFragment =>
        lookupCachedData(currentFragment)
          .map(_.cachedRepresentation.withOutput(currentFragment.output))
          .getOrElse(currentFragment)
    }
    
    newPlan transformAllExpressions {
      case s: SubqueryExpression => s.withNewPlan(useCachedData(s.plan))
    }
  }
```